src/librustc/ty/layout.rs

```diff
@@ -549,23 +561,37 @@ impl<'a, 'gcx, 'tcx> Struct {
             min_size: Size::from_bytes(0),
         };
 
-        // 1-member and 2-member structs don't optimize.
+        if fields.len() == 0 {return Ok(ret)};
+
+        // Anything with ReprExtern or ReprPacked doesn't optimize.
+        // Neither do  1-member and 2-member structs.
         // In addition, code in trans assume that 2-element structs can become pairs.
         // It's easier to just short-circuit here.
-        let can_optimize_struct = fields.len() > 2;
+        let mut can_optimize = fields.len() > 2 || StructKind::EnumVariant == kind;
+        if can_optimize {
+            // This exhaustive match makes new reprs force the adder to modify this function.
+            // Otherwise, things can silently break.
+            // Note the inversion, return true to stop matching.
+            can_optimize = !reprs.iter().any(|r| {
+                match *r {
+                    attr::ReprAny => false,
+                    attr::ReprInt(_) => false,
+                    attr::ReprExtern => true,
+                    attr::ReprPacked => true,
+                    attr::ReprSimd => bug!("Simd  vectors should be represented as a layout::Vector")
+                }
+            });
+        }
 
-        let (optimize, sort_ascending) = match (repr, kind) {
-            (attr::ReprAny, StructKind::AlwaysSizedUnivariant) => (can_optimize_struct, false),
-            (attr::ReprAny, StructKind::MaybeUnsizedUnivariant) => (can_optimize_struct, true),
-            (attr::ReprAny, StructKind::EnumVariant) => {
+        let (optimize, sort_ascending) = match kind {
+            StructKind::AlwaysSizedUnivariant => (can_optimize, false),
+            StructKind::MaybeUnsizedUnivariant => (can_optimize, true),
+            StructKind::EnumVariant => {
                 assert!(fields.len() >= 1, "Enum variants must have discriminants.");
-                (true, fields[0].size(dl).bytes() == 1)
+                (can_optimize, fields[0].size(dl).bytes() == 1)
```

Comment:
The two pair values here appears to be inverted. That is, `fields[0].size(dl).bytes() == 1` is about optimization, while `sort_ascending` is `true` for enums.